server/ServerApp.cpp

```diff
@@ -965,6 +966,52 @@ namespace {
                                    << "human player with assigned empire id: " << psd.m_save_game_empire_id;
         }
     }
+
+    /** Replace species selection for any players set to "RANDOM" species, with a species based on @p initial_seed and player id.
+     *  If changed, seed is reset to @p initial_seed before return. */
+    void ResolveRandomSpeciesSelections(std::map<int, PlayerSetupData>& player_setup_data, const unsigned int initial_seed) {
+        SpeciesManager& sm = GetSpeciesManager();
+        int num_playable_species = sm.NumPlayableSpecies();
+        // early return for no playable species
+        if (num_playable_species <= 0) {
+            ErrorLogger() << "No playable species found";
+            return;
+        }
+        bool seed_altered = false;
+
+        // evenly initialize a list for weighting each species
+        float prob_init_value = 1.0f / num_playable_species;
+        std::vector<float> species_prob;
+        for (int i = 0; i < num_playable_species; ++i)
+            species_prob.push_back(prob_init_value);
+
+        float weight_reduction = prob_init_value / num_playable_species;
+
+        // iterate over players
+        for (std::map<int, PlayerSetupData>::value_type& entry : player_setup_data) {
+            if (entry.second.m_starting_species_name == "RANDOM") {
+                int player_id = entry.first;
+                // alter the seed for each player according to their id
+                seed_altered = true;
+                Seed(initial_seed + player_id);
+                unsigned int seed = initial_seed * RandGaussian(1.0, (player_id * 0.01));
+                Seed(seed);
+                int idx = RandWeightedInt(species_prob);
+
+                // reduce the chance of this species being picked again
+                species_prob[idx] -= weight_reduction;
```

Comment:
You probably do not want to do this. 

Let n be the number of species and m the number of times a species is picked.

Then, your chance to pick the species m+1 times is ```p(m+1) = p(m) - 1/n^2 = 1/n - m/n^2```. This obviously runs into trouble if m > n.  This **will** happen if the total number of random selection exceeds n^2. (Consider only 1 playable species but 3 players that want a random selection).
Not sure if the boost function you call has some fancy stuff happening to handle weights that sum to zero or even negative probabilites but you leave mathematical definitions behind in any case.

Instead, why not reduce probability by a set factor each time the species is selects, e.g. ```species_prob[idx] *= .5```, i.e. ```p(m+1) = p(m) * 0.5 = p(initial) * 0.5^m```.
where m is number of times this species is picked. Seems more intuitive and does also not run into the potential problems above.